rpc/personal_api.go

```diff
@@ -1,30 +1,168 @@
+	params "github.com/cosmos/ethermint/rpc/args"
+	"github.com/spf13/viper"
 
+	"github.com/ethereum/go-ethereum/accounts"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
 // PersonalEthAPI is the eth_ prefixed set of APIs in the Web3 JSON-RPC spec.
 type PersonalEthAPI struct {
-	cliCtx    sdkcontext.CLIContext
-	nonceLock *AddrLocker
+	cliCtx      sdkcontext.CLIContext
+	ethAPI      *PublicEthAPI
+	nonceLock   *AddrLocker
+	keys        []emintcrypto.PrivKeySecp256k1
+	keybaseLock sync.Mutex
 }
 
 // NewPersonalEthAPI creates an instance of the public ETH Web3 API.
-func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, nonceLock *AddrLocker) *PersonalEthAPI {
+func NewPersonalEthAPI(cliCtx sdkcontext.CLIContext, ethAPI *PublicEthAPI, nonceLock *AddrLocker, keys []emintcrypto.PrivKeySecp256k1) *PersonalEthAPI {
 	return &PersonalEthAPI{
 		cliCtx:    cliCtx,
+		ethAPI:    ethAPI,
 		nonceLock: nonceLock,
+		keys:      keys,
 	}
 }
 
+func (e *PersonalEthAPI) getKeybaseInfo() ([]keyring.Info, error) {
+	e.keybaseLock.Lock()
+	defer e.keybaseLock.Unlock()
+
+	if e.cliCtx.Keybase == nil {
+		keybase, err := keyring.NewKeyring(
+			sdk.KeyringServiceName(),
+			viper.GetString(flags.FlagKeyringBackend),
+			viper.GetString(flags.FlagHome),
+			e.cliCtx.Input,
+			keyring.WithKeygenFunc(emintcrypto.EthermintKeygenFunc),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		e.cliCtx.Keybase = keybase
+	}
+
+	return e.cliCtx.Keybase.List()
+}
+
+// ImportRawKey stores the given hex encoded ECDSA key into the key directory,
+// encrypting it with the passphrase.
+func (e *PersonalEthAPI) ImportRawKey(privkey, password string) (common.Address, error) {
+	return common.Address{}, nil
+}
+
+// ListAccounts will return a list of addresses for accounts this node manages.
+func (e *PersonalEthAPI) ListAccounts() ([]common.Address, error) {
+	addrs := []common.Address{}
+	infos, err := e.getKeybaseInfo()
+	if err != nil {
+		return addrs, err
+	}
+
+	for _, info := range infos {
+		addressBytes := info.GetPubKey().Address().Bytes()
+		addrs = append(addrs, common.BytesToAddress(addressBytes))
+	}
+
+	return addrs, nil
+}
+
+// LockAccount will lock the account associated with the given address when it's unlocked.
+func (e *PersonalEthAPI) LockAccount(address common.Address) bool {
+	for i, key := range e.keys {
+		if bytes.Equal(key.PubKey().Address().Bytes(), address.Bytes()) {
+			tmp := make([]emintcrypto.PrivKeySecp256k1, len(e.keys)-1)
+			copy(tmp[:i], e.keys[:i])
+			copy(tmp[i:], e.keys[i+1:])
+			e.keys = tmp
+			return true
+		}
+	}
+
+	return false
+}
+
+// NewAccount will create a new account and returns the address for the new account.
+func (e *PersonalEthAPI) NewAccount(password string) (common.Address, error) {
+	_, err := e.getKeybaseInfo()
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	// TODO: is keyring.Secp256k1 the correct algorithm?
```

Comment:
hmmm yeah, but we should define ours by passing in the `WithSupportedAlgo` keybase option when creating it. If you pass in the `keyring.Secp256k1` it will create a Tendermint secp251k1 key

Reply:
will update once #439  is merged